frontend/src/hocs/withScreencast/withScreencast.tsx

```diff
@@ -0,0 +1,267 @@
+ * limitations under the License.
+ */
+
+import classNames from "classnames"
+import WhichBrowser from "which-browser"
+import ScreenCastRecorder from "lib/ScreenCastRecorder"
+import hoistNonReactStatics from "hoist-non-react-statics"
+import React, { PureComponent, ComponentType, ReactNode } from "react"
+
+import {
+  UnsupportedBrowserDialog,
+  ScreencastDialog,
+  VideoRecordedDialog,
+} from "hocs/withScreencast/components"
+
+import "./withScreencast.scss"
+
+interface WithScreenCastProps {}
+
+interface WithScreenCastState {
+  fileName: string
+  recording: boolean
+  recordAudio: boolean
+  countdown: number
+  startAnimation: boolean
+  blob?: Blob
+  showRecordedDialog: boolean
+  showScreencastDialog: boolean
+  showUnsupportedDialog: boolean
+}
+
+export interface ScreenCastHOC extends WithScreenCastState {
+  toggleRecordAudio: () => void
+  startRecording: (fileName: string) => void
+  stopRecording: () => void
+}
+
+const SUPPORTED_BROWSERS: any = {
+  Chrome: "49",
+  Firefox: "29",
+  Opera: "36",
+  Edge: "76",
+}
+
+function withScreencast(
+  WrappedComponent: ComponentType<any>
+): ComponentType<any> {
+  class ComponentWithScreencast extends PureComponent<
+    WithScreenCastProps,
+    WithScreenCastState
+  > {
+    static readonly displayName = `withScreencast(${WrappedComponent.displayName ||
+      WrappedComponent.name})`
+
+    recorder?: ScreenCastRecorder | null
+
+    state = {
+      fileName: "streamlit-screencast",
+      countdown: -1,
+      recording: false,
+      recordAudio: false,
+      startAnimation: false,
+      showRecordedDialog: false,
+      showScreencastDialog: false,
+      showUnsupportedDialog: false,
+    }
+
+    toggleRecordAudio = (): void => {
+      const { recordAudio } = this.state
+
+      this.setState({
+        recordAudio: !recordAudio,
+      })
+    }
+
+    showDialog = (fileName: string): void => {
+      const { recording, countdown } = this.state
+
+      if (!this.checkSupportedBrowser()) {
+        this.setState({
+          showUnsupportedDialog: true,
+        })
+      } else {
+        if (!recording && countdown < 0) {
+          this.setState({
+            fileName,
+            showScreencastDialog: true,
+          })
+        } else {
+          this.stopRecording()
+        }
+      }
+    }
+
+    startRecording = async (): Promise<any> => {
+      const { recordAudio } = this.state
+      this.recorder = new ScreenCastRecorder({ recordAudio })
+
+      await this.recorder.initialize()
+
+      this.setState({
+        countdown: 3,
+        startAnimation: true,
+        showRecordedDialog: false,
+      })
+    }
+
+    stopRecording = async (): Promise<any> => {
+      let blob
+      const { recording } = this.state
+
+      if (
+        this.recorder &&
+        recording &&
+        this.recorder.getState() !== "inactive"
+      )
+        blob = await this.recorder.stop()
+
+      this.setState({
+        blob,
+        countdown: -1,
```

Comment:
Does the value of `countdown` matter in this case?

Reply:
I'm just resetting the complete state to the initial one.